### ISSUE TITLE
[BUG] JS string comparison raises an exception

### ISSUE THREAD

**Issue by monstasat** [ocsigen/js_of_ocaml#895]
**Describe the bug**
I've noticed that code like `Dom_html.document##.readyState = Js.string "loading"` throws an exception. Exception has the following message: `compare: abstract value`.
Here is the stacktrace from my production code:

```caml_raise_with_arg | @ | stdlib.js:117
  | caml_raise_with_string | @ | stdlib.js:122
  | caml_invalid_argument | @ | stdlib.js:178
  | caml_compare_val | @ | stdlib.js:424
  | caml_equal | @ | stdlib.js:455
  | domContentLoaded | @ | lwt_js_events.ml:537
```
It seems that the code fails because of the `caml_compare_val` function, in the following [line](https://github.com/ocsigen/js_of_ocaml/blob/0052b39e079a64d6886a29c4e66a41b4ea9e63c0/runtime/stdlib.js#L424).
The `Lwt_js_events.domContentLoaded` function which calls `caml_equal` just compares `Dom_html.document##.readyState` with some constant at line 537.

Here is a another simple test illustrating the issue:
```ocaml
(* This is ok and shows [true] in console *)
print_endline "comparing two equal js strings";
print_endline (string_of_bool (Js.string "foo" = Js.string "foo"));
(* This is ok and shows [false] in console *)
print_endline "comparing two different js strings";
print_endline (string_of_bool (Js.string "loading" = Js.string "complete"));
(* This one fails*)
let state = Dom_html.document##.readyState in
print_endline "comparing state";
print_endline (string_of_bool (state = Js.string "complete"))
```

**Expected behavior**
This should not fail.

**Versions**
jsoo commit 0052b39e079a64d6886a29c4e66a41b4ea9e63c0

**Comment by hhugo** [comment]
Thanks for the report, this was introduced by https://github.com/ocsigen/js_of_ocaml/pull/889.
I think @TyOverby is looking at a fix.

**Comment by monstasat** [comment]
Seems that string issue was fixed by #896.
I've also faced similar problem with list comparison.
I have a [react](https://github.com/dbuenzli/react) signal of type `Dom_html.element Js.t list` with implicit polymorphic compare. When this signal changes, an exception with `compare: abstract value` message is raised.

**Comment by hhugo** [comment]
What semantic do you want/expect when comparing `Dom_html.element Js.t` ?

**Comment by monstasat** [comment]
I expect that DOM elements are equal if they are the same object (both have equal reference).
I personally use physical equality for them (`==`).

**Comment by monstasat** [comment]
This problem may obviously occur in many places where implicit polymorphic compare takes place, e.g. `List.mem`, `React.signal` and so on. Maybe it is better to use physical equality when unknown semantic is met rather than raising an exception? The least I expect from the compare function is throwing an exception (except functional compare).

**Comment by dbuenzli** [comment]
I was first suspicious about @monstasat proposal but what he suggests seems to be compatible with OCaml's view on its own objects (`( = )` is true iff the two objects are physically equal).

**Comment by hhugo** [comment]
@monstasat, can you give another try at #896 ?

**Comment by hhugo** [comment]
Fixed by #903
